### PR TITLE
Atomic rate limiter implementation

### PR DESCRIPTION
**Pull request by storozhukBM**
Hi,
I've made simple atomic implementation of your RateLimiter (actually it's Go port of original Java implementation from [Resilience4j](https://github.com/resilience4j/resilience4j))
This implementation is a little bit harder but it is generally faster and has no goroutine scalability issues  that any mutex based implementation have. 
Checkout this relation between RateLimiter latency and count of goroutines using this RateLimiter:

comparison

Original implementation moved to mutex_rate.go file, so you can run benchmarks and play with other tests. But I'd suggest to leave only one implementation after all.

You can also checkout my benchmark results here [AtomicRLvsMutexRL.pdf](https://github.com/beefsack/go-rate/files/1640752/AtomicRLvsMutexRL.pdf)

### CODE REVIEW

**Comment by storozhukBM** [comment]
Seems like benchmarks code is not supported in old Go versions, but we can actually remove it before merge.

File: mutex_rate.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package rate
```
Comment (beefsack) [inline mutex_rate.go]: If we have a new implementation, I see no need to keep the `Mutex` version.

File: rate_test.go [hunk 2]
```diff
@@ -3,8 +3,90 @@ package rate
...
+			}
+
+			b.StartTimer()
+			wg.Wait()
+			b.StopTimer()
+		})
+	}
+}
+
+func BenchmarkMutexRateLimiter(b *testing.B) {
```
Comment (beefsack) [inline rate_test.go]: If we remove the `Mutex` version we will need to remove this benchmark too.

File: rate.go [hunk 3]
```diff
@@ -44,18 +56,43 @@ func (r *RateLimiter) Wait() {
...
 // remaining time before the rate limit expires.
 func (r *RateLimiter) Try() (ok bool, remaining time.Duration) {
-	r.mtx.Lock()
-	defer r.mtx.Unlock()
-	now := time.Now()
-	if l := r.times.Len(); l < r.limit {
-		r.times.PushBack(now)
-		return true, 0
+	for {
+		//extract previous state
```
Comment (beefsack) [inline rate.go]: Space after `//`

File: rate.go [hunk 4]
```diff
@@ -44,18 +56,43 @@ func (r *RateLimiter) Wait() {
...
+		// count elapsed cycles and produce more permissions if necessary
+		elapsedCycles := currentCycle - previousState.cycle
+		if elapsedCycles > 0 {
+			permissionsToAppear := elapsedCycles * r.limit
+			newState.permissions = min(previousState.permissions+permissionsToAppear, r.limit)
+		}
+		// try to acquire permission by atomic update
+		if newState.permissions > 0 {
+			newState.permissions -= 1
+			if atomic.CompareAndSwapPointer((*unsafe.Pointer)(&r.state), previousStatePointer, unsafe.Pointer(&newState)) {
```
Comment (beefsack) [inline rate.go]: Long line, can we have args on newlines?

File: rate.go [hunk 5]
```diff
@@ -44,18 +56,43 @@ func (r *RateLimiter) Wait() {
...
+				return true, 0
+			}
+			continue
+		}
+		// if there is not enough permissions calculate wait duration
+		nextCycleStart := (currentCycle + 1) * int64(r.interval)
```
Comment (beefsack) [inline rate.go]: Unnecessary cast.

File: rate.go [hunk 6]
```diff
@@ -44,18 +56,43 @@ func (r *RateLimiter) Wait() {
...
+		nsToNextCycle := nextCycleStart - now
+		fullCyclesRequired := (-newState.permissions) / r.interval
+		remaining := (fullCyclesRequired * r.interval) + nsToNextCycle
```
Comment (beefsack) [inline rate.go]: Unnecessary parens.

File: rate.go [hunk 7]
```diff
@@ -44,18 +56,43 @@ func (r *RateLimiter) Wait() {
...
+			newState.permissions -= 1
+			if atomic.CompareAndSwapPointer((*unsafe.Pointer)(&r.state), previousStatePointer, unsafe.Pointer(&newState)) {
+				return true, 0
+			}
+			continue
+		}
+		// if there is not enough permissions calculate wait duration
+		nextCycleStart := (currentCycle + 1) * int64(r.interval)
+		nsToNextCycle := nextCycleStart - now
+		fullCyclesRequired := (-newState.permissions) / r.interval
```
Comment (beefsack) [inline rate.go]: Unnecessary parens.

**Comment by storozhukBM** [comment]
Hi @beefsack, thank you for this feedback. I want to assure you that I've made my best to achieve thread safety for this implementation. Off-course, I've used race detector during tests of this implementation, but it is not an 100% guarantee. As always when you face hard concurrency problem the only tool you can use it's your mind and understanding of memory model. The only gotcha with usage of atomics in GO is [this](https://golang.org/src/sync/atomic/doc.go?s=1206:1656#L36).
But we are safe here, because our rate limiter state have proper 64-bit alignment.

Also I want to ask for explicit permission to use this implementation in other open source libraries that are not compatible with GNU General Public License v3.0
Or, maybe if we are changing implementation completely, we can change license to MIT or Hi @beefsack, thank you for this feedback. I want to assure you that I've made my best to achieve thread safety for this implementation. Off-course, I've used race detector during tests of this implementation, but it is not an 100% guarantee. As always when you face hard concurrency problem the only tool you can use it's your mind and understanding of memory model. The only gotcha with usage of atomics in GO is [this](https://golang.org/src/sync/atomic/doc.go?s=1206:1656#L36).
But we are safe here, because our rate limiter state have proper 64-bit alignment.

Also I need to ask for explicit permission to use this implementation in other open source libraries that are not compatible with GNU General Public License v3.0
Or, maybe if we are changing implementation completely, we can change license to "MIT" or "Apache License 2.0", but this is up to you off-course.

**Comment by storozhukBM** [comment]
I'm closing this PR due to lack of response.

**Pull request closed without merging**